### ISSUE TITLE
Reloading ontology should not cause Protege to hang

### ISSUE THREAD

**Issue by rsgoncalves** [protegeproject/protege#475]
After changing an ontology file outside of Protege (while the same ontology was loaded in Protege), and switching back to Protege, it prompts to reload the ontology for disk. After accepting to reload, the following appears in the console, but Protege comes to a complete halt.

    ------------------------------ Reloading ontology ------------------------------ 
    Reloading ontology:
    OntologyID(OntologyIRI(<http://www.semanticweb.org/rgoncalves/ontologies/2016/6/untitled-ontology-120-ok>) VersionIRI(<null>)) 
    Copied ontologies in 0 ms.  Used: 0 MB 
    Applying 7005 change(s) to patch ontology to reloaded ontology 

Thread dump:

    2016-07-07 10:33:49
    Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.74-b02 mixed mode):
    
    "pool-4-thread-1" #29 prio=5 os_prio=31 tid=0x00007fee89564800 nid=0x12267 waiting on condition [0x000070000f488000]
       java.lang.Thread.State: WAITING (parking)
    	at sun.misc.Unsafe.park(Native Method)
    	- parking to wait for  <0x0000000540d25308> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
    	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
    	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
    	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
    	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    	at java.lang.Thread.run(Thread.java:745)
    
    "pool-2-thread-1" #28 prio=5 os_prio=31 tid=0x00007fee89ac8000 nid=0x115cf waiting on condition [0x000070001148e000]
       java.lang.Thread.State: WAITING (parking)
    	at sun.misc.Unsafe.park(Native Method)
    	- parking to wait for  <0x0000000540806908> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
    	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
    	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
    	at java.util.concurrent.LinkedBlockingQueue.take(LinkedBlockingQueue.java:442)
    	at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1067)
    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1127)
    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    	at java.lang.Thread.run(Thread.java:745)
    
    "TimerQueue" #24 daemon prio=5 os_prio=31 tid=0x00007fee8b4ae800 nid=0x3e1f waiting on condition [0x000070001048b000]
       java.lang.Thread.State: WAITING (parking)
    	at sun.misc.Unsafe.park(Native Method)
    	- parking to wait for  <0x00000005406c5db8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
    	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
    	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2039)
    	at java.util.concurrent.DelayQueue.take(DelayQueue.java:211)
    	at javax.swing.TimerQueue.run(TimerQueue.java:171)
    	at java.lang.Thread.run(Thread.java:745)
    
    "Timer-0" #20 daemon prio=5 os_prio=31 tid=0x00007fee8a63d000 nid=0x10a07 in Object.wait() [0x000070000e485000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at java.lang.Object.wait(Object.java:502)
    	at java.util.TimerThread.mainLoop(Timer.java:526)
    	- locked <0x000000054002bc28> (a java.util.TaskQueue)
    	at java.util.TimerThread.run(Timer.java:505)
    
    "DestroyJavaVM" #19 prio=5 os_prio=31 tid=0x00007fee8c831000 nid=0x1703 waiting on condition [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "OSGi Shutdown Thread" #18 prio=5 os_prio=31 tid=0x00007fee8c84a000 nid=0x1080f in Object.wait() [0x000070000d482000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	- waiting on <0x000000054003bbb8> (a org.apache.felix.framework.util.ThreadGate)
    	at org.apache.felix.framework.util.ThreadGate.await(ThreadGate.java:79)
    	- locked <0x000000054003bbb8> (a org.apache.felix.framework.util.ThreadGate)
    	at org.apache.felix.framework.Felix.waitForStop(Felix.java:1006)
    	at org.protege.osgi.framework.Launcher.lambda$addCleanupOnExit$1(Launcher.java:146)
    	at org.protege.osgi.framework.Launcher$$Lambda$19/1330754528.run(Unknown Source)
    	at java.lang.Thread.run(Thread.java:745)
    
    "AWT-EventQueue-0" #16 prio=6 os_prio=31 tid=0x00007fee8b2d5000 nid=0xef07 runnable [0x000070000c3f7000]
       java.lang.Thread.State: RUNNABLE
    	at java.lang.Object.notifyAll(Native Method)
    	at javax.swing.text.AbstractDocument.writeUnlock(AbstractDocument.java:1364)
    	- locked <0x0000000540bfcf40> (a javax.swing.text.DefaultStyledDocument)
    	at javax.swing.text.DefaultStyledDocument.setParagraphAttributes(DefaultStyledDocument.java:583)
    	at org.protege.editor.owl.ui.renderer.OWLCellRenderer.resetStyles(OWLCellRenderer.java:929)
    	at org.protege.editor.owl.ui.renderer.OWLCellRenderer.prepareTextPane(OWLCellRenderer.java:664)
    	at org.protege.editor.owl.ui.renderer.OWLCellRenderer.prepareRenderer(OWLCellRenderer.java:483)
    	at org.protege.editor.owl.ui.renderer.OWLCellRenderer.getListCellRendererComponent(OWLCellRenderer.java:392)
    	at javax.swing.plaf.basic.BasicListUI.updateLayoutState(BasicListUI.java:1361)
    	at javax.swing.plaf.basic.BasicListUI.maybeUpdateLayoutState(BasicListUI.java:1311)
    	at javax.swing.plaf.basic.BasicListUI$Handler.valueChanged(BasicListUI.java:2623)
    	at javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:184)
    	at javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:164)
    	at javax.swing.DefaultListSelectionModel.fireValueChanged(DefaultListSelectionModel.java:211)
    	at javax.swing.DefaultListSelectionModel.insertIndexInterval(DefaultListSelectionModel.java:632)
    	at javax.swing.plaf.basic.BasicListUI$Handler.intervalAdded(BasicListUI.java:2581)
    	at javax.swing.AbstractListModel.fireIntervalAdded(AbstractListModel.java:148)
    	at javax.swing.DefaultComboBoxModel.insertElementAt(DefaultComboBoxModel.java:138)
    	at org.protege.editor.search.ui.SortedComboBoxModel.insertElementAt(SortedComboBoxModel.java:57)
    	at org.protege.editor.search.ui.SortedComboBoxModel.addElement(SortedComboBoxModel.java:35)
    	at org.protege.editor.search.ui.OwlEntityComboBox.addItem(OwlEntityComboBox.java:68)
    	at org.protege.editor.search.ui.OwlEntityComboBoxChangeHandler.visit(OwlEntityComboBoxChangeHandler.java:36)
    	at org.semanticweb.owlapi.model.AddAxiom.accept(AddAxiom.java:71)
    	at org.protege.editor.search.ui.BasicQueryPanel.lambda$new$0(BasicQueryPanel.java:83)
    	at org.protege.editor.search.ui.BasicQueryPanel$$Lambda$98/359326989.ontologiesChanged(Unknown Source)
    	at org.semanticweb.owlapi.model.DefaultChangeBroadcastStrategy.broadcastChanges(DefaultChangeBroadcastStrategy.java:38)
    	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.broadcastChanges(OWLOntologyManagerImpl.java:1490)
    	at uk.ac.manchester.cs.owl.owlapi.OWLOntologyManagerImpl.applyChanges(OWLOntologyManagerImpl.java:590)
    	at org.protege.editor.owl.model.OntologyReloader.reload(OntologyReloader.java:60)
    	at org.protege.editor.owl.model.OWLModelManagerImpl.reload(OWLModelManagerImpl.java:395)
    	at org.protege.editor.owl.ui.ontology.OntologySourcesChangedHandlerUI.ontologySourcesChanged(OntologySourcesChangedHandlerUI.java:104)
    	at org.protege.editor.owl.model.io.OntologySourcesManager.checkSources(OntologySourcesManager.java:58)
    	at org.protege.editor.owl.ui.ontology.OntologySourcesChangedHandlerUI.handleWindowActivated(OntologySourcesChangedHandlerUI.java:170)
    	at org.protege.editor.owl.ui.ontology.OntologySourcesChangedHandlerUI$1$1.windowActivated(OntologySourcesChangedHandlerUI.java:62)
    	at java.awt.AWTEventMulticaster.windowActivated(AWTEventMulticaster.java:390)
    	at java.awt.AWTEventMulticaster.windowActivated(AWTEventMulticaster.java:389)
    	at java.awt.AWTEventMulticaster.windowActivated(AWTEventMulticaster.java:389)
    	at java.awt.Window.processWindowEvent(Window.java:2070)
    	at javax.swing.JFrame.processWindowEvent(JFrame.java:305)
    	at java.awt.Window.processEvent(Window.java:2017)
    	at java.awt.Component.dispatchEventImpl(Component.java:4891)
    	at java.awt.Container.dispatchEventImpl(Container.java:2294)
    	at java.awt.Window.dispatchEventImpl(Window.java:2750)
    	at java.awt.Component.dispatchEvent(Component.java:4713)
    	at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
    	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:995)
    	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:488)
    	at java.awt.Component.dispatchEventImpl(Component.java:4762)
    	at java.awt.Container.dispatchEventImpl(Container.java:2294)
    	at java.awt.Window.dispatchEventImpl(Window.java:2750)
    	at java.awt.Component.dispatchEvent(Component.java:4713)
    	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
    	at java.awt.EventQueue.access$500(EventQueue.java:97)
    	at java.awt.EventQueue$3.run(EventQueue.java:709)
    	at java.awt.EventQueue$3.run(EventQueue.java:703)
    	at java.security.AccessController.doPrivileged(Native Method)
    	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
    	at java.awt.EventQueue$4.run(EventQueue.java:731)
    	at java.awt.EventQueue$4.run(EventQueue.java:729)
    	at java.security.AccessController.doPrivileged(Native Method)
    	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
    	at java.awt.SentEvent.dispatch(SentEvent.java:70)
    	at java.awt.DefaultKeyboardFocusManager$DefaultKeyboardFocusManagerSentEvent.dispatch(DefaultKeyboardFocusManager.java:217)
    	at java.awt.DefaultKeyboardFocusManager.sendMessage(DefaultKeyboardFocusManager.java:244)
    	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:366)
    	at java.awt.Component.dispatchEventImpl(Component.java:4762)
    	at java.awt.Container.dispatchEventImpl(Container.java:2294)
    	at java.awt.Window.dispatchEventImpl(Window.java:2750)
    	at java.awt.Component.dispatchEvent(Component.java:4713)
    	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
    	at java.awt.EventQueue.access$500(EventQueue.java:97)
    	at java.awt.EventQueue$3.run(EventQueue.java:709)
    	at java.awt.EventQueue$3.run(EventQueue.java:703)
    	at java.security.AccessController.doPrivileged(Native Method)
    	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
    	at java.awt.EventQueue$4.run(EventQueue.java:731)
    	at java.awt.EventQueue$4.run(EventQueue.java:729)
    	at java.security.AccessController.doPrivileged(Native Method)
    	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
    	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
    	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
    	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
    	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
    	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
    	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
    	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
    
    "Java2D Disposer" #15 daemon prio=10 os_prio=31 tid=0x00007fee89332800 nid=0xdb13 in Object.wait() [0x000070000b3f9000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
    	- locked <0x0000000540025d78> (a java.lang.ref.ReferenceQueue$Lock)
    	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
    	at sun.java2d.Disposer.run(Disposer.java:148)
    	at java.lang.Thread.run(Thread.java:745)
    
    "Java2D Queue Flusher" #14 daemon prio=10 os_prio=31 tid=0x00007fee892ef000 nid=0xdd07 in Object.wait() [0x000070000a3f6000]
       java.lang.Thread.State: TIMED_WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at sun.java2d.opengl.OGLRenderQueue$QueueFlusher.run(OGLRenderQueue.java:203)
    	- locked <0x000000054001d6b8> (a sun.java2d.opengl.OGLRenderQueue$QueueFlusher)
    
    "AWT-Shutdown" #13 prio=5 os_prio=31 tid=0x00007fee892d0000 nid=0x8413 in Object.wait() [0x0000700009164000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at java.lang.Object.wait(Object.java:502)
    	at sun.awt.AWTAutoShutdown.run(AWTAutoShutdown.java:295)
    	- locked <0x000000054003be58> (a java.lang.Object)
    	at java.lang.Thread.run(Thread.java:745)
    
    "AppKit Thread" #12 daemon prio=5 os_prio=31 tid=0x00007fee892d3000 nid=0xa0b runnable [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "FelixStartLevel" #11 daemon prio=5 os_prio=31 tid=0x00007fee8a18f000 nid=0x5d03 in Object.wait() [0x000070000805b000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	- waiting on <0x000000054002bf68> (a java.util.ArrayList)
    	at java.lang.Object.wait(Object.java:502)
    	at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:279)
    	- locked <0x000000054002bf68> (a java.util.ArrayList)
    	at java.lang.Thread.run(Thread.java:745)
    
    "FelixDispatchQueue" #10 prio=5 os_prio=31 tid=0x00007fee8b12f800 nid=0x5b0f in Object.wait() [0x0000700007058000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at java.lang.Object.wait(Object.java:502)
    	at org.apache.felix.framework.util.EventDispatcher.run(EventDispatcher.java:1074)
    	- locked <0x000000054003c720> (a java.util.ArrayList)
    	at org.apache.felix.framework.util.EventDispatcher.access$000(EventDispatcher.java:55)
    	at org.apache.felix.framework.util.EventDispatcher$1.run(EventDispatcher.java:102)
    	at java.lang.Thread.run(Thread.java:745)
    
    "Service Thread" #9 daemon prio=9 os_prio=31 tid=0x00007fee8b00b800 nid=0x5303 runnable [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "C1 CompilerThread3" #8 daemon prio=9 os_prio=31 tid=0x00007fee8984b800 nid=0x5103 waiting on condition [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "C2 CompilerThread2" #7 daemon prio=9 os_prio=31 tid=0x00007fee89004000 nid=0x4f03 waiting on condition [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "C2 CompilerThread1" #6 daemon prio=9 os_prio=31 tid=0x00007fee8b00e000 nid=0x4d03 waiting on condition [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "C2 CompilerThread0" #5 daemon prio=9 os_prio=31 tid=0x00007fee8a05f800 nid=0x4b03 waiting on condition [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "Signal Dispatcher" #4 daemon prio=9 os_prio=31 tid=0x00007fee8a05e800 nid=0x3a13 waiting on condition [0x0000000000000000]
       java.lang.Thread.State: RUNNABLE
    
    "Finalizer" #3 daemon prio=8 os_prio=31 tid=0x00007fee8b00a800 nid=0x3803 in Object.wait() [0x0000700003a3a000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
    	- locked <0x0000000540043070> (a java.lang.ref.ReferenceQueue$Lock)
    	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
    	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)
    
    "Reference Handler" #2 daemon prio=10 os_prio=31 tid=0x00007fee8b008000 nid=0x3603 in Object.wait() [0x0000700002a37000]
       java.lang.Thread.State: WAITING (on object monitor)
    	at java.lang.Object.wait(Native Method)
    	at java.lang.Object.wait(Object.java:502)
    	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
    	- locked <0x0000000540043228> (a java.lang.ref.Reference$Lock)
    	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)
    
    "VM Thread" os_prio=31 tid=0x00007fee8a03f000 nid=0x3403 runnable 
    
    "GC task thread#0 (ParallelGC)" os_prio=31 tid=0x00007fee8b003800 nid=0x2403 runnable 
    
    "GC task thread#1 (ParallelGC)" os_prio=31 tid=0x00007fee8a001000 nid=0x2603 runnable 
    
    "GC task thread#2 (ParallelGC)" os_prio=31 tid=0x00007fee8a003000 nid=0x2803 runnable 
    
    "GC task thread#3 (ParallelGC)" os_prio=31 tid=0x00007fee89018800 nid=0x2a03 runnable 
    
    "GC task thread#4 (ParallelGC)" os_prio=31 tid=0x00007fee89019000 nid=0x2c03 runnable 
    
    "GC task thread#5 (ParallelGC)" os_prio=31 tid=0x00007fee8901a000 nid=0x2e03 runnable 
    
    "GC task thread#6 (ParallelGC)" os_prio=31 tid=0x00007fee8901a800 nid=0x3003 runnable 
    
    "GC task thread#7 (ParallelGC)" os_prio=31 tid=0x00007fee8901b000 nid=0x3203 runnable 
    
    "VM Periodic Task Thread" os_prio=31 tid=0x00007fee8901b800 nid=0x5503 waiting on condition 
    
    JNI global references: 1867
    
    Heap
     PSYoungGen      total 3445760K, used 100955K [0x00000006eab00000, 0x00000007c0000000, 0x00000007c0000000)
      eden space 3410944K, 2% used [0x00000006eab00000,0x00000006eed9fdf0,0x00000007bae00000)
      from space 34816K, 94% used [0x00000007bde00000,0x00000007bfdf7008,0x00000007c0000000)
      to   space 41984K, 0% used [0x00000007bae00000,0x00000007bae00000,0x00000007bd700000)
     ParOldGen       total 494080K, used 330421K [0x0000000540000000, 0x000000055e280000, 0x00000006eab00000)
      object space 494080K, 66% used [0x0000000540000000,0x00000005542ad7d8,0x000000055e280000)
     Metaspace       used 40192K, capacity 41624K, committed 41896K, reserved 1085440K
      class space    used 5545K, capacity 5949K, committed 6056K, reserved 1048576K

**Comment by matthewhorridge** [comment]
@rsgoncalves do you have the file by any chance?  Also, was it in a folder with a lot of other files or sub-files?

**Comment by matthewhorridge** [comment]
Closing because I cannot reproduce this

**Comment by rsgoncalves** [comment]
I tracked down the ontology file, but it's changed since then. It was definitely in a folder with a lot of other files.

**Comment by matthewhorridge** [comment]
o.k. thanks for the update.  It's very likely a duplicate of https://github.com/protegeproject/protege/issues/83

Can you confirm that when you say it's changed you mean the problem doesn't happen anymore?

**Comment by rsgoncalves** [comment]
Could well be, as the file was in a folder with ~70 ontology files, and subfolders with more ontologies. 

No, I meant the ontology with which I encountered this issue has gone through axiom changes since then.

**Comment by matthewhorridge** [comment]
o.k. thanks.  I'd say it's definitely https://github.com/protegeproject/protege/issues/83.
